src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/PreferDictionaryTryAddValueOverGuardedAddAnalyzer.cs

```diff
@@ -0,0 +1,149 @@
+            {
+                var additionalLocations = ImmutableArray.Create(dictionaryAddLocation);
+                context.ReportDiagnostic(Diagnostic.Create(DiagnosticDescriptor, i.Syntax.GetLocation(), additionalLocations));
+            }
+        }
+
+        private static bool TryGetEligibleDictionaryAddLocation(IOperation? conditionalBody, IOperation dictionaryInstance, IOperation containsKeyArgument, IMethodSymbol addSymbol, [NotNullWhen(true)] out Location? dictionaryAddLocation)
+        {
+            dictionaryAddLocation = null;
+            if (conditionalBody is null)
+            {
+                return false;
+            }
+
+            static bool OperationEquals(SemanticModel semanticModel, IOperation left, IOperation right)
+            {
+                var leftSymbol = semanticModel.GetSymbolInfo(left.Syntax).Symbol;
+                var rightSymbol = semanticModel.GetSymbolInfo(right.Syntax).Symbol;
+
+                return SymbolEqualityComparer.Default.Equals(leftSymbol, rightSymbol);
+            }
+
+            bool IsEligibleDictionaryAddInvocation(IInvocationOperation invocation)
+            {
+                return OperationEquals(invocation.SemanticModel, invocation.Instance, dictionaryInstance)
+                       && DoesSignatureMatch(invocation.TargetMethod, addSymbol)
+                       && invocation.Arguments[0].Value.Syntax.IsEquivalentTo(containsKeyArgument.Syntax)
+                       && invocation.Arguments[1].Value.Kind == OperationKind.Literal;
+            }
+
+            if (conditionalBody is IBlockOperation block)
+            {
+                dictionaryAddLocation = block.Operations.OfType<IExpressionStatementOperation>().FirstOrDefault(e => e.Operation is IInvocationOperation i && IsEligibleDictionaryAddInvocation(i))?.Operation.Syntax.GetLocation();
+            }
+            else if (conditionalBody is IExpressionStatementOperation { Operation: IInvocationOperation iOperation } && IsEligibleDictionaryAddInvocation(iOperation))
+            {
+                dictionaryAddLocation = iOperation.Syntax.GetLocation();
+            }
+
+            return dictionaryAddLocation is not null;
+        }
+
+        private static bool TryGetDictionaryTypeAndMembers(
+            Compilation compilation,
+            [NotNullWhen(true)] out INamedTypeSymbol? iDictionaryType,
+            [NotNullWhen(true)] out IMethodSymbol? containsKeySymbol,
+            [NotNullWhen(true)] out IMethodSymbol? addSymbol)
+        {
+            containsKeySymbol = null;
+            addSymbol = null;
+            iDictionaryType = WellKnownTypeProvider.GetOrCreate(compilation).GetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemCollectionsGenericIDictionary2);
+            if (iDictionaryType is null)
+            {
+                return false;
+            }
```

Comment:
Does it work like that?

```suggestion
            iDictionaryType = WellKnownTypeProvider.GetOrCreate(compilation).GetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemCollectionsGenericIDictionary2);
            if (iDictionaryType is null)
            {
                containsKeySymbol = null;
                addSymbol = null;
                return false;
            }
```

Reply:
Yup